src/SpacetimeDBClient.cs

```diff
@@ -377,30 +388,66 @@ HashSet<byte[]> GetInsertHashSet(System.Type tableType, int tableSize)
                         {
                             Log.Warn("Non-insert during a subscription update!");
                         }
-
                         foreach (var bin in BsatnRowListIter(qu.Inserts))
                         {
                             if (!hashSet.Add(bin))
                             {
                                 // Ignore duplicate inserts in the same subscription update.
                                 continue;
                             }
-
                             var obj = table.DecodeValue(bin);
                             var op = new DbOp
                             {
                                 table = table,
                                 insert = new(obj, bin),
                             };
-
                             dbOps.Add(op);
                         }
                     }
                 }
-
                 return (dbOps, subscriptionInserts);
             }
 
+            /// <summary>
+            /// TODO: the dictionary is here for backwards compatibility and can be removed
+            /// once we get rid of legacy subscriptions.
+            /// </summary>
+            (List<DbOp>, Dictionary<System.Type, HashSet<byte[]>>) PreProcessSubscribeApplied(SubscribeApplied subscribeApplied)
+            {
+                var table = clientDB.GetTable(subscribeApplied.Rows.TableName) ?? throw new Exception($"Unknown table name: {subscribeApplied.Rows.TableName}");
+                var dbOps = new List<DbOp>();
+                HashSet<byte[]> inserts = new();
+
+                // First apply all of the state
+                foreach (var cqu in subscribeApplied.Rows.TableRows.Updates)
+                {
+                    var qu = DecompressDecodeQueryUpdate(cqu);
+                    if (qu.Deletes.RowsData.Count > 0)
+                    {
+                        Log.Warn("Non-insert during a subscription update!");
+                    }
+                    foreach (var bin in BsatnRowListIter(qu.Inserts))
+                    {
+                        if (!inserts.Add(bin))
+                        {
+                            // Ignore duplicate inserts in the same subscription update.
+                            continue;
+                        }
+                        var obj = table.DecodeValue(bin);
+                        var op = new DbOp
+                        {
+                            table = table,
+                            insert = new(obj, bin),
+                        };
+                        dbOps.Add(op);
+                    }
+                }
```

Comment:
Looks like this entire loop is duplicated with PreProcessLegacySubscription. I know we're going to eventually remove the legacy one, but for now can it be extracted into a helper function to simplify maintainance?